rpcs3/Emu/Cell/Modules/sceNp.cpp

```diff
@@ -609,14 +608,68 @@ error_code sceNpDrmExecuteGamePurchase()
 
 error_code sceNpDrmGetTimelimit(vm::cptr<char> path, vm::ptr<u64> time_remain)
 {
-	sceNp.todo("sceNpDrmGetTimelimit(path=%s, time_remain=*0x%x)", path, time_remain);
+	sceNp.warning("sceNpDrmGetTimelimit(path=%s, time_remain=*0x%x)", path, time_remain);
 
 	if (!path || !time_remain)
 	{
 		return SCE_NP_DRM_ERROR_INVALID_PARAM;
 	}
 
-	*time_remain = SCE_NP_DRM_TIME_INFO_ENDLESS;
+	//*time_remain = SCE_NP_DRM_TIME_INFO_ENDLESS;
+
+	vm::var<s64> sec;
+	vm::var<s64> nsec;
+
+	// Get system time (real or fake) to compare to
+	error_code ret = sys_time_get_current_time(sec, nsec);
+	if (ret != CELL_OK)
+		return ret;
+
+	const std::string enc_drm_path(path.get_ptr(), std::find(path.get_ptr(), path.get_ptr() + 0x100, '\0'));
+	const auto [fs_error, ppath, real_path, enc_file, type] = lv2_file::open(enc_drm_path, 0, 0);
+
+	if (fs_error)
+	{
+		return {fs_error, enc_drm_path};
+	}
+
+	u32 magic;
+	NPD_HEADER npd;
+
+	enc_file.read<u32>(magic);
+	enc_file.seek(0);
+
+	// Read expiration time from NPD header which is Unix timestamp in milliseconds
+	if (magic == "SCE\0"_u32)
+	{
+		if (!get_npdrm_self_header(enc_file, npd))
+		{
+			sceNp.error(u8"sceNpDrmGetTimelimit(): Failed to read NPD header from sce file '%s'", enc_drm_path);
+			return {SCE_NP_DRM_ERROR_INVALID_PARAM, enc_drm_path};
+		}
+	}
+	else if (magic == "NPD\0"_u32)
+	{
+		// edata / sdata files
+		EDAT_HEADER edat;
+		read_npd_edat_header(&enc_file, npd, edat);
+	}
+	else
+	{
+		// Unknown file type
+		return {SCE_NP_DRM_ERROR_INVALID_PARAM, enc_drm_path};
+	}
+
+	// Convert milliseconds to seconds
+	s64 expire_time = npd.expire_time / 1000ll;
```

Comment:
```suggestion
	const s64 expire_time = npd.expire_time / 1000ll;
```

Reply:
Actually, why do you remove accuracy here?
The result of the function is in microseconds, so it would make more sense to keep the exact expire_time and calculate the result with a factor of 1000...

Reply:
Right, I change the whole calculation to use milliseconds but I don't want to replace cellRtcGetTickResolution() with a magic number. Any ideas?

Reply:
I don't see a reason to use the function at all.
It's clear that it returns microseconds.
You are already basing your calculations on seconds, which is an arbitrary "magic" choice in the first place